libs/core/langchain_core/indexes/base.py

```diff
@@ -0,0 +1,369 @@
+from abc import ABC, abstractmethod
+from typing import (
+    Any,
+    AsyncIterable,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+    cast,
+)
+
+from langchain_core.documents import Document
+from langchain_core.indexes.types import DeleteResponse, UpsertResponse
+from langchain_core.runnables import run_in_executor
+from langchain_core.stores import BaseStore
+from langchain_core.structured_query import StructuredQuery
+
+
+class Index(BaseStore[str, Document], ABC):
+    """Interface for a document index.
+
+    Example:
+        .. code-block:: python
+
+            from typing import Any, Iterator, List, Optional, Sequence, Tuple, Union, Iterable
+            from uuid import uuid4
+
+            from langchain_core.documents import Document
+            from langchain_core.indexes import UpsertResponse, DeleteResponse, Index
+
+            def uuid4_generator() -> Iterable[str]:
+                while True:
+                    yield str(uuid4())
+
+            class DictIndex(Index):
+
+                def __init__(self) -> None:
+                    self.store = {}
+
+                def upsert(
+                    self,
+                    documents: Iterable[Document],
+                    *,
+                    ids: Optional[Iterable[str]] = None,
+                    **kwargs: Any,
+                ) -> UpsertResponse:
+                    ids = ids or uuid4_generator()
+                    succeeded = []
+                    for id_, doc in zip(ids, documents):
+                        self.store[id_] = doc
+                        succeeded.append(id_)
+                    return UpsertResponse(succeeded=succeeded, failed=[])
+
+                def delete_by_ids(self, ids: Iterable[str]) -> DeleteResponse:
+                    succeeded = []
+                    failed = []
+                    for id_ in ids:
+                        try:
+                            del self.store[id_]
+                        except Exception:
+                            failed.append(id_)
+                        else:
+                            succeeded.append(id_)
+                    return DeleteResponse(succeeded=succeeded, failed=failed)
+
+                def lazy_get_by_ids(self, ids: Iterable[str]) -> Iterable[Document]:
+                    for id in ids:
+                        yield self.store[id]
+
+                def yield_keys(
+                    self, *, prefix: Optional[str] = None
+                ) -> Union[Iterator[str]]:
+                    prefix = prefix or ""
+                    for key in self.store:
+                        if key.startswith(prefix):
+                            yield key
+    """  # noqa: E501
+
+    @abstractmethod
+    def upsert(
+        self,
+        # TODO: Iterable or Iterator?
+        documents: Iterable[Document],
+        *,
+        ids: Optional[Iterable[str]] = None,
+        **kwargs: Any,
+    ) -> UpsertResponse:
```

Comment:
If we're making the input an Iterable, then the output should be an iterator over UpsertResponse